### ISSUE TITLE
Linux: Investigate how to configure BB so it can send documents from LibreOffice or from a folder/deskop via "Sent To > Mail Recient"

### ISSUE THREAD

**Issue by Betterbird** [Betterbird/thunderbird-patches#85]
On newsgroup de.comm.software.mozilla.mailnews some discussion was started that TB 102.3.0 couldn't receive documents from LO to be sent to a mail recipient. However, various users have since reported that it does in fact work.

It doesn't appear to work for BB since there's some configuration missing. We're hoping to get some feedback from @Posi81 or @mfschumann as to what may be missing. Does it work in the Flatpak?

**Comment by mfschumann** [comment]
I have done a quick test with the BB flatpak 102.3.1-bb18 under Gnome 42 on Fedora Silverblue 36.

When I try to send a document from LO a BB compose window is opened with the subject set to the filename of the document. What seems to be missing is that the document is attached to the mail. For the TB flatpak the behavior is identical. When calling `flatpak run eu.betterbird.Betterbird -compose "attachment='/var/home/martin/tmp/somefilethatexists'"` I get a warning that the file does not exist (although it does), so maybe the issue is related to the flatpak sandboxing. I'll try to debug.

**Comment by Betterbird** [comment]
Partial success:
Putting a `betterbird.desktop` into `/usr/share/applications` and editing `/usr/share/applications/defaults.list` to contain `x-scheme-handler/mailto=betterbird.desktop` has the effect that mailto links in FF now open in BB. I guess that was the easy part. The question remains how LO and `Sent To > Mail Recipient` have it configured.

**Comment by Betterbird** [comment]
Using `ps aux | grep thunder` I saw that when attaching a document from LO while TB is the default client (via the Mint/Xfce "Preferred Applications" panel), this command is run:
`/usr/lib/thunderbird/thunderbird -remote "xfeDoCommand(composeMessage,to='',cc='',bcc='',attachment='file:///home/xxxxx/Desktop/betterbird/application.ini',subject='noname')"`
No idea where this is configured, however supported in the TB code here:
https://searchfox.org/comm-central/rev/bfefc713c577d1b7db4e06f16bd687ccb7875ea7/mail/components/MessengerContentHandler.jsm#256
That appears to be starting TB 78 on our old Mint 20 VM.

Running an equivalent command for BB, like:
`/home/xxxxx/Desktop/betterbird102/betterbird -remote "xfeDoCommand(composeMessage,to='',cc='',bcc='',attachment='file:///home/xxxxx/Desktop/betterbird/application.ini',subject='noname')"`
launches a compose window without the attachment.

Your attempts with `-compose "attachment='/var/home/martin/tmp/somefilethatexists'"` don't appear to reflect what LO calls. If we just configure `betterbird -compose "%s"` as default mail app, we see this command being executed:
`betterbird -compose ?subject=noname&attach=file:///tmp/lu41567qh31by.tmp/lu41567qh31dx.tmp/noname.odt`
Obviously the URL-query notation is not compatible with the `-compose` parameters, hence the integration via `-remote` instead. Also, trying to add an attachment via a mailto: URL (like the `?subject=`) would be a security hole and doesn't work.

**Comment by Betterbird** [comment]
Looks like there is a script that changes one format to another. We have it at `/usr/share/xfce4/exo/exo-compose-mail`. Looks like this one here: https://www.apt-browse.org/browse/ubuntu/xenial/universe/amd64/libexo-1-0/0.10.7-1/file/usr/lib/x86_64-linux-gnu/xfce4/exo-1/exo-compose-mail-1

**Comment by Betterbird** [comment]
Summarising: On an old Mint system with TB 78 the following all works:
```
/usr/lib/thunderbird/thunderbird -remote "xfeDoCommand(composeMessage,attachment='file:///home/jorgk/Desktop/huhu.txt',subject='noname')"
/usr/lib/thunderbird/thunderbird -compose "attachment='file:///home/jorgk/Desktop/huhu.txt',subject=noname)"
betterbird -compose "attachment='file:///home/jorgk/Desktop/huhu.txt',subject=noname)"
```

What doesn't work is
```
betterbird -remote "xfeDoCommand(composeMessage,attachment='file:///home/jorgk/Desktop/huhu.txt',subject='noname')"
```
That also doesn't work on Windows.

**Comment by mfschumann** [comment]
Even if you get the `betterbird -remote` call working from command line, for the integration with LO to work I think we additionally need to file a PR with LO, because they use a [script that distinguishes between mail clients](https://github.com/LibreOffice/core/blob/master/shell/source/unix/misc/senddoc.sh) and has some special treatment for Thunderbird.

**Comment by Betterbird** [comment]
Thanks for digging this out. Fixing LO is a one liner, no? `-remote` is broken in TB as well (of course): https://bugzilla.mozilla.org/show_bug.cgi?id=1793323. So somehow the people who claim it's working here
https://www.thunderbird-mail.de/forum/thread/90511-thunderbird-102-geht-verschicken-eines-pdf-dokumentes-aus-libreoffice-oder-aus-e/?postID=504181#post504181
are triggering the code path where LO uses `-compose`.

**Comment by mfschumann** [comment]
> When calling `flatpak run eu.betterbird.Betterbird -compose "attachment='/var/home/martin/tmp/somefilethatexists'"` I get a warning that the file does not exist (although it does), so maybe the issue is related to the flatpak sandboxing. I'll try to debug.

I found that indeed the error I got was caused by lacking permissions in the sandbox that the flatpak app runs in. According to https://github.com/flathub/org.mozilla.Thunderbird/issues/263 the preferred solution to that issue (using 'Portals' for file access) would have to be implemented upstream in the TB code. At the moment it seems to be partially implemented, leading to incosistent behavior that I documented in [the readme for the BB flatpak](https://github.com/flathub/eu.betterbird.Betterbird/blob/master/README.md#lacking-file-permissions).

There is a work-around (giving BB access to /home), but applying that comes with the caveat that the user has to move its profile directory. So I am hesitant to generally apply that work-around for the published flatpak.

All in all, for the "Send mail" integration between LO and BB to work for the flatpak version, quite a list of things have to happen:
- LO adds Betterbird to the list of mail agents that are treated like Thunderbird
- TB/BB upstream repairs the functionality of the `-remote "xfeDoCommand(composeMessage,attachment=...` call
- TB/BB upstream fixes the file permission issue described above by using 'Portals' everywhere

**Comment by Betterbird** [comment]
1. https://bugs.documentfoundation.org/show_bug.cgi?id=151314 - Beyond our control
2. https://bugzilla.mozilla.org/show_bug.cgi?id=1793323 - Will get fixed in the next BB release, either TB or we will fix.
3. https://bugzilla.mozilla.org/show_bug.cgi?id=1786251 - Hmm, likely no quick fix, but there's a workaround.

We've only played with Mint/Xfce's "Preferred Programs" panel. That could be used to launch a BB compose window from a very old version of LO. We'll have to see what happens with Nr, 2 fixed even without Nr. 1.

**Comment by Betterbird** [comment]
TDF people suggested to provide a patch for Nr. 1. @mfschumann could you patch their senddoc.sh script and test it once a fix for Nr. 2 is available? Off-topic: You are one of the reporters of [bug 799040](https://bugzilla.mozilla.org/show_bug.cgi?id=799040#c13) (Issues with ignored threads), well, that was fixed in BB a while ago with TB still not having picked up the patch.

**Comment by mfschumann** [comment]
Thanks, I will test your bugfix and try to patch LO's senddoc.sh locally. This may take some time because it seems that the script lands in the compiled binary, so I will need to set up the build first.

**Comment by mfschumann** [comment]
This is my understanding of the code paths involved:
![image](https://user-images.githubusercontent.com/9102962/194613141-5b6def95-770e-4d4b-af01-21cf14890c10.png)

Your fix repairs the path where you end up in 5a, i.e., the user must manually set the mail client in LO. On the LO side this requires `senddoc.sh` to include `betterbird` in the list of Thunderbird-like clients. I've implemented that change locally, rebuilt the LO flatpak and confirmed that it calls betterbird Thunderbird-style instead via `xdg-email`. I will add that patch to the [TDF issue](https://bugs.documentfoundation.org/show_bug.cgi?id=151314).

For the code path ending in 5c, i.e., when the user does not specify the mail client within LO settings, there will probably be no fix because support of attaching files via `mailto` arguments is considered unsafe by TB upstream. Adding to that, there is also [a proposal to drop support for attaching files in `xdg-email` ](https://gitlab.freedesktop.org/xdg/xdg-utils/-/merge_requests/58).

For the flatpak versions of LO and BB the issue is even more complex due to the application sandboxing getting in the way when trying to call one app from within the other app's sandbox and sharing files between apps. Making the integration between flatpak'ed LO and BB work very probably requires either some manual work-arounds on the user side or changes to the LO code. Plus the result will be brittle because [flatpak lacks a stable API to realize stuff like this](https://github.com/flatpak/xdg-desktop-portal/issues/283). So I am not very inclined to push this topic any further.

**Comment by Betterbird** [comment]
Impressive investigation. However, the findings don't explain what I see in my old LO Writer 6.4.7.2 on Mint 20/Xfce, likely things have changed since then:
1. No mail client configured in LO, Thunderbird configured as default maill app: Mail is sent via TB with the attachment, so we're not in 5c, in fact, via `ps aux | grep thunder` I can see 5a.
2. BB configured in LO, Thunderbird configured as default maill app: Behaves the same as 1, although according to your diagram we should also get to 5c.
3. BB configured as default mail app (regardless of setting in LO): Nothing happens.

But never mind, I have a very old version of LO and I don't know what the default e-mail configuration or xdg-mail do.

As far as LO is concerned, your patch in https://bugs.documentfoundation.org/show_bug.cgi?id=151314 looks good to me.

**Comment by fitojb** [comment]
@mfschumann I’ve merged your [patch](https://github.com/LibreOffice/core/commit/42fe257fa4d970f58543ca40f5158c04f607bc61) and will propose it for inclusion in the upcoming LO 7.4.3. Thanks!

**Comment by mfschumann** [comment]
With the fix implemented, I was able to develop a work-around for the flatpak-version of Betterbird as documented [here](https://github.com/flathub/eu.betterbird.Betterbird/blob/develop/README.md#sending-documents-as-mail-attachments-from-libreoffice).

**Comment by Betterbird** [comment]
Hi @mfschumann, it's still working right? We got some reports here:
https://www.reddit.com/r/Betterbird/comments/1so32bj/comment/oguqspx/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

**Comment by mfschumann** [comment]
Yes, the work-around documented [here](https://github.com/flathub/eu.betterbird.Betterbird/blob/develop/README.md#sending-documents-as-mail-attachments-from-libreoffice) for flatpak BB with flatpak LO still works for me.

I haven't tested the "flatpak BB with non-flatpak LO" variant, but I am confident that it should work, as it's one less layer of complexity than the both-flatpak variant.

**Comment by Betterbird** [comment]
Unfortunately, my Ubuntu 22 VM has LibreOffice 7.3.7.2 on it, so it's too old to test since the LO fix came out in 7.4.3.

**Comment by Betterbird** [comment]
Well, the user comments further on Reddit:

> BUT the setting in LO seems to be ignored in favor of whatever setting I have for the Preferred Application in the Linux Mint GUI. When the Linux Mint default email program is set to Betterbird, an open Betterbird message appears with 0 attachments. When I flip the default email program to Thunderbird, LO opens a Thunderbird message with the file attached.

**Comment by mfschumann** [comment]
You could suggest running `ps aux | grep betterbird` when the Betterbird window has opened after clicking 'send via mail', so we could see in which of [the known code paths](https://github.com/Betterbird/thunderbird-patches/issues/85#issuecomment-1271865427) the user ends up. Maybe that helps to narrow down the issue.

Are they using Cinnamon or any other desktop environment?

**Comment by tedfranklin** [comment]
I'm just returning to this issue ... Yes, I'm using Cinnamon on Linux Mint.  When I flip back and forth between Betterbird and Thunderbird in LM's Preferred Applications app (and change nothing else), the result is that LibreOffice Send by Email works perfectly with Thunderbird preferred but, when Betterbird is the preferred email app LO fires up a Betterbird compose window but does not attach the document (i.e., the subject line correctly reflects the title of the document).  Picking up on @mfschumann's suggestion, here is what i see when I run ps aux | grep betterbird:

``` ps aux | grep betterbird
ted       626270  0.0  0.0   3584  2076 ?        S    12:39   0:00 bwrap --args 70 -- betterbird mailto:?subject=test&attach=file:///home/ted/Downloads/LOtmp/lu375701g7fo.tmp/lu375701g7ly.tmp/test.odt&attachment=file:///home/ted/Downloads/LOtmp/lu375701g7fo.tmp/lu375701g7ly.tmp/test.odt&attach=/home/ted/Downloads/LOtmp/lu375701g7fo.tmp/lu375701g7ly.tmp/test.odt
ted       626283  0.1  0.0   3584  1568 ?        S    12:39   0:00 bwrap --args 70 -- betterbird mailto:?subject=test&attach=file:///home/ted/Downloads/LOtmp/lu375701g7fo.tmp/lu375701g7ly.tmp/test.odt&attachment=file:///home/ted/Downloads/LOtmp/lu375701g7fo.tmp/lu375701g7ly.tmp/test.odt&attach=/home/ted/Downloads/LOtmp/lu375701g7fo.tmp/lu375701g7ly.tmp/test.odt
ted       626284 39.7  1.7 3376084 564980 ?      Sl   12:39   0:05 /app/lib/betterbird/betterbird mailto:?subject=test&attach=file:///home/ted/Downloads/LOtmp/lu375701g7fo.tmp/lu375701g7ly.tmp/test.odt&attachment=file:///home/ted/Downloads/LOtmp/lu375701g7fo.tmp/lu375701g7ly.tmp/test.odt&attach=/home/ted/Downloads/LOtmp/lu375701g7fo.tmp/lu375701g7ly.tmp/test.odt
ted       626394  0.8  0.3 2468508 109028 ?      Sl   12:39   0:00 /app/lib/betterbird/betterbird-bin -contentproc -isForBrowser -prefsHandle 0:7759 -prefMapHandle 1:280705 -jsInitHandle 2:242716 -parentBuildID 20260520100101 -sandboxReporter 3 -ipcHandle 4 -initialChannelId {eb6b120c-d2f1-45ee-b40b-6e538f0b7a7e} -parentPid 2 -greomni /app/lib/betterbird/omni.ja -appDir /app/lib/betterbird 1 tab
ted       626396  0.2  0.1 362936 51584 ?        Sl   12:39   0:00 /app/lib/betterbird/betterbird-bin -contentproc -parentBuildID 20260520100101 -prefsHandle 0:7759 -prefMapHandle 1:280705 -sandboxReporter 2 -ipcHandle 3 -initialChannelId {daa346b4-4808-45f2-a024-bd1635591033} -parentPid 2 -appDir /app/lib/betterbird 2 rdd
ted       626430  0.9  0.3 2495444 117136 ?      Sl   12:39   0:00 /app/lib/betterbird/betterbird-bin -contentproc -isForBrowser -prefsHandle 0:4933 -prefMapHandle 1:280705 -jsInitHandle 2:242716 -parentBuildID 20260520100101 -sandboxReporter 3 -ipcHandle 4 -initialChannelId {786b78cc-8d53-4e2b-b122-0efe6ba28688} -parentPid 2 -greomni /app/lib/betterbird/omni.ja -appDir /app/lib/betterbird 3 tab
ted       626687  0.0  0.0   9144  2264 pts/0    S+   12:39   0:00 grep --color=auto betterbird```

and here is what I get when I run ps aux | grep thunderbird:
``` ps aux | grep thunderbird
ted       629293  5.9  1.0 11506044 327124 ?     Sl   12:42   0:01 /usr/lib/thunderbird/thunderbird -compose subject=test,attachment='file:///home/ted/Downloads/LOtmp/lu375701g7fo.tmp/lu375701g7lz.tmp/test.odt,/home/ted/Downloads/LOtmp/lu375701g7fo.tmp/lu375701g7lz.tmp/test.odt'
ted       629299  0.0  0.0  21080  2812 ?        Sl   12:42   0:00 /usr/lib/thunderbird/crashhelper 629293 8 /tmp/ 10
ted       629874  0.0  0.0   9144  2272 pts/0    S+   12:43   0:00 grep --color=auto thunderbird```

I am not able to decipher this, but I can see that a lot is going wrong when I'm sending from Betterbird.  I spent too much time with Claude yesterday trying to understand sensible-lomua, senddoc, xdg-email, etc.

**Comment by mfschumann** [comment]
Thanks for the `ps aux` output. It seems that Betterbird is being called via `xdg-email mailto` (which is broken because it does not support attachments) while thunderbird is called directly with the `-compose` argument. This discrimination should actually have been fixed in LibreOffice 7.4.3 four years ago. Are you using an older version of LibreOffice?

**Comment by tedfranklin** [comment]
I'm using Version: 25.8.7.3 (X86_64) / LibreOffice Community Build ID: 580(Build:3) having been dealt this particular fork when I installed Linux Mint.  Am I forking down the wrong fork? I came back to this problem because I had no luck with the "official" solution posted on the Betterbird website.  Some changes to xdg-email fix the problem (at long last) for me.  However, if the version of LibreOffice currently being distributed with Linux Mint does not contain code that already fixed the problem,  I'm concerned that I may be causing unnecessary confusion with my bug report here:  https://gitlab.freedesktop.org/xdg/xdg-utils/-/work_items/298 but I was stuck in the mud and found a way out for my setup.  I believe you have a much better idea than I of how this all works.  If I should pull my bug report, let me know. Thanks!
